0000-interledger-protocol-4/0000-interledger-protocol-4.md

```diff
@@ -0,0 +1,146 @@
+| `message` | UTF8 String | User-readable error message, primarily intended for debugging purposes |
+| `data` | Variable-Length Octet String | End-to-end data. Connectors MUST NOT modify this data |
+
+See the ASN.1 definition [here](../asn1/InterledgerProtocol.asn).
+
+### Error Codes
+
+#### F__ - Final Error
+
+Final errors indicate that the payment is invalid and should not be retried unless the details are changed.
+
+| Code | Name | Description | Data Fields |
+|---|---|---|---|
+| **F00** | **Bad Request** | Generic sender error. | (empty) |
+| **F01** | **Invalid Packet** | The ILP packet was syntactically invalid. | (empty) |
+| **F02** | **Unreachable** | There was no way to forward the payment, because the destination ILP address was wrong or the connector does not have a route to the destination. | (empty) |
+| **F03** | **Invalid Amount** | The amount is invalid, for example it contains more digits of precision than are available on the destination ledger or the amount is greater than the total amount of the given asset in existence. | (empty) |
+| **F04** | **Insufficient Destination Amount** | The receiver deemed the amount insufficient, for example you tried to pay a $100 invoice with $10. | (empty) |
+| **F05** | **Wrong Condition** | The receiver generated a different condition and cannot fulfill the payment. | (empty) |
+| **F06** | **Unexpected Payment** | The receiver was not expecting a payment like this (the memo and destination address don't make sense in that combination, for example if the receiver does not understand the transport protocol used) | (empty) |
+| **F07** | **Cannot Receive** | The receiver is unable to accept this payment due to a constraint. For example, the payment would put the receiver above its maximum account balance. | (empty) |
+| **F99** | **Application Error** | Reserved for application layer protocols. Applications MAY use names other than `Application Error`. | Determined by Application |
+
+#### T__ - Temporary Error
+
+Temporary errors indicate a failure on the part of the receiver or an intermediary system that is unexpected or likely to be resolved soon. Senders SHOULD retry the same payment again, possibly after a short delay.
+
+| Code | Name | Description | Data Fields |
+|---|---|---|---|
+| **T00** | **Internal Error** | A generic unexpected exception. This usually indicates a bug or unhandled error case. | (empty) |
+| **T01** | **Ledger Unreachable** | The connector has a route or partial route to the destination but was unable to reach the next ledger. Try again later. | (empty) |
+| **T02** | **Ledger Busy** | The ledger is rejecting requests due to overloading. Try again later. | (empty) |
+| **T03** | **Connector Busy** | The connector is rejecting requests due to overloading. Try again later. | (empty) |
+| **T04** | **Insufficient Liquidity** | The connector would like to fulfill your request, but it doesn't currently have enough money. Try again later. | (empty) |
```

Comment:
Is "liquidity" or "money" here the same as the "bandwidth" described in the Flow section?

Reply:
Yes. I think the main difference between the bandwidth and balance concepts is that bandwidth captures the fact that it may refill quickly if you have a fast settlement method like a payment channel